### ISSUE TITLE
Ensemble training score gets worse with time

### ISSUE THREAD

**Issue by CYHSM** [automl/auto-sklearn#106]
I noticed for my dataset that after around 10000s of training the training score was 0.57 and i had the maximum time on 18000s. The final score it gave me was 0.54, which i think is odd because he was finding an ensemble before with a better score. 
This is my code:

```
automl = autosklearn.classification.AutoSklearnClassifier(
        #ml_memory_limit=6500,
        time_left_for_this_task=time_left_for_this_task,
        per_run_time_limit=per_run_time_limit,
        delete_tmp_folder_after_terminate=True,
        delete_output_folder_after_terminate=False,
        resampling_strategy='cv',
        resampling_strategy_arguments={'folds': 10},
        output_folder=output_dir,
        tmp_folder=tmp_dir,
        include_estimators=['gradient_boosting'])


automl.fit(train_inputs.copy(), train_outputs.copy())
automl.refit(train_inputs.copy(), train_outputs.copy())
```

**Comment by mfeurer** [comment]
How do you assess the final score? Is it the last training score the ensemble outputs or do you obtain it with automl.score()?

**Comment by M-OOC-F** [comment]
First i check the training score in the output and there its getting worse and then also afterwards i check with `automl.ensemble_.training_score` 

I think i cant use automl.score because it does not return the cross validated accuracy, right?

**Comment by mfeurer** [comment]
No, you can't. automl.score requires you to pass X and y, so it's there for getting the test performance.

Regarding the problem: the ensemble building process is a greedy strategy similar. It selects the model which increases the training score most in each iteration and adds it to the ensemble. If the best model changes, this can change the whole ensemble. For that it's hard to tell whether it harms your performance. 

Figuring this out would require you to build each ensemble created during the optimization process on the whole dataset. Then you could compare the training and testing error. 

A different solution could be to add a technique called Bagged Ensemble Selection to auto-sklearn. It should make the ensemble creation less prone to changes due to the best model changing.

**Comment by M-OOC-F** [comment]
Ok that makes sense, but why is then the model not verified versus the ensemble. So ensemble before vs. ensemble + new classifier. 
As i understand its currently : best classifier from ensemble vs. current classifier and then its added to the ensemble if theres a better one?

**Comment by mfeurer** [comment]
We're using a technique called ensemble selection: http://www.cs.cornell.edu/~caruana/ctp/ct.papers/caruana.icml04.icdm06long.pdf with the addition of what's described in Section 2.1. Ensemble Selection is started from scratch in each iteration. We could only decide to keep the old ensemble if we had a seperate test set. Bagged Ensemble Selection which I mentioned earlier is described in Section 2.3.

**Comment by mfeurer** [comment]
I'm closing this now because your question seem to be answered. If this is not the case, please reopen this issue.
